core/src/test/groovy/com/novoda/gradle/release/AndroidDifferentGradleVersions.groovy

```diff
@@ -0,0 +1,54 @@
+package com.novoda.gradle.release
+
+import com.novoda.gradle.release.rule.TestProjectRule
+import org.gradle.testkit.runner.GradleRunner
+import org.gradle.testkit.runner.TaskOutcome
+import org.junit.Rule
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.junit.runners.Parameterized
+
+import static org.assertj.core.api.Assertions.assertThat
+
+@RunWith(Parameterized.class)
+class AndroidDifferentGradleVersions {
+
+    @Rule
+    public TestProjectRule projectRule = new TestProjectRule(TestProjectRule.Project.ANDROID)
+
+    @Parameterized.Parameters(name = "{index}: test Gradle version {0}")
+    static Collection<GradleVerionsParams> gradleVersionExpectedOutcome() {
+        return [
+                // Gradle 4.0 is not support by the Android Gradle Plugin 3.x
+                new GradleVerionsParams("4.0", true),
+                new GradleVerionsParams("4.1", TaskOutcome.SUCCESS),
+                new GradleVerionsParams("4.2", TaskOutcome.SUCCESS),
+                new GradleVerionsParams("4.3", TaskOutcome.SUCCESS),
+                new GradleVerionsParams("4.4", TaskOutcome.SUCCESS),
+                // TODO: Failure on Gradle 4.5. is **not** expected. It's failing because of changes in
+                // the UsageContext in our AndroidLibrary class
+                new GradleVerionsParams("4.5", true),
+        ]
+    }
+
+    private GradleVerionsParams testParams
+
+    AndroidDifferentGradleVersions(GradleVerionsParams testParams) {
+        this.testParams = testParams
+    }
+
+    @Test
+    void testDifferentGradleVersionsAndOutcome() {
```

Comment:
what is the outcome you are testing? 

`testOurUseOfInternalAPisWorksWithDifferentGradleVersions` ?

Reply:
I don't just test the internal APIs only. I test the whole project.
Your test name sounds like "If we don't use internals anymore we can remove the test". Which I don't suggest.

Since we are using a Parameterized test I don't want to add "works" here because not all version works. And works would be the wrong outcome. Maybe we expect that it doesn't work (like with Gradle 4.0).

So I'm happy with my current method name.

Reply:
the name I suggested was an example of being explicit with describing the expectations.

Right now "outcome" isn't clear what is expected. If you are struggling to name it descriptively you could alternatively add javadoc, for instance with what you wrote above.

side note: this is a problem i find with parameterized tests. They may be "cool" at "speeding up" writing tests - but long term for legacy code developer supporters - they are confusing.

Reply:
What about this

`givenGradleVersion_WhenProjectBuild_ShouldHaveExpectedOutcome`

Reply:
@blundell I suggested `Parameterized` test. The reason is definitely not being cool or speeding up. It is more about maintainability. 

This class already has a single purpose. That's why I don't think it will not be confusing. This just tests if the build has expected outcome. We are not expecting any additional tests here.

If you do X different tests for X different cases here, you will have lots of duplication in the test code and also in test names. There is also no guarantee that future developers will keep them consistent.

Reply:
"`ShouldHaveExpectedOutcome`" is the problem

_ExpectedOutcome_ needs defining

Reply:
To be that is obvious. The params are not even primitives. `GradleVerionsParams` is introduced to encapsulate the `gradle versions` and `outcome`.

Reply:
Please come up with a suggestion to make the test naming more clear.

Reply:
```
                // Gradle 4.0 is not support by the Android Gradle Plugin 3.x
 +                new GradleVerionsParams("4.0", true),
 +                new GradleVerionsParams("4.1", TaskOutcome.SUCCESS),
 +                new GradleVerionsParams("4.2", TaskOutcome.SUCCESS),
 +                new GradleVerionsParams("4.3", TaskOutcome.SUCCESS),
 +                new GradleVerionsParams("4.4", TaskOutcome.SUCCESS),
 +                // TODO: Failure on Gradle 4.5. is **not** expected. It's failing because of changes in
 +                // the UsageContext in our AndroidLibrary class
 +                new GradleVerionsParams("4.5", true),
```

true vs success is not obvious

`testWeOnlySupportSpecificGradleVersions`

Reply:
> "ShouldHaveExpectedOutcome" is the problem

I'm with you that "Outcome" is not a good naming here. 
With testing names we should have something like `calcularAdd_2plus5_shouldReturn7`.
The seven here is the expected output.
We should have someting like ` assertTrue(result, 7)`.
Mapping this to our test it is equivalent. Because we check if the `TaskOutcome` have a specific value (which we define in the Parameter method).

I think the name "Outcome" is here the problem?! But since the Gradle class `TaskOutcome` sounds outcome it is totally fine IMO.

`testWeOnlySupportSpecificGradleVersions` 
is also not that clear. Because we test ALSO the TaskOutput.
If we don't check the output with the `assertThat` but only the `build()` or `buildFail()` from the GradleRunner I would agree with that name. But as we also check the TaskOutput the method name removes some information ... :)

> true vs success is not obvious

I'm with you. It is a little bit crazy. In a previous implementation I had only the three params constructor. But since the `TaskOutcome` don't care if we already fail with `buildAndFail()`.
I can try to improve it with a named parameter... http://docs.groovy-lang.org/latest/html/documentation/#_named_argument_constructor

Reply:
In a parameterized test, the output already has the name of the parameters for each run. This PR  will make it say 
```
test Gradle version: GradleVerionsParams{ gradleVersion='4.4.1', expectedGradleBuildFailure=false, expectedTaskOutcome=SUCCESS }
```

Given that, the test name should be about that particular param, not generic.  Considering that `testWeOnlySupportSpecificGradleVersions` will not be suitable because it does not really reflect what the test does. 

To be honest, I don't think the test name is super important because the param names in the test run makes it super clear.

Reply:
so whats the next steps?